File: nhs_backend/java/src/main/java/com/bhtask/nhs/backend/service/ElasticsearchConditionsServiceImpl.java
package com.bhtask.nhs.backend.service;

import static org.elasticsearch.index.query.QueryBuilders.*;

import java.io.BufferedReader;
import java.io.File;
import java.io.FileReader;
import java.util.ArrayList;
import java.util.List;

import javax.annotation.PostConstruct;

import org.elasticsearch.index.query.QueryBuilder;
import org.slf4j.Logger;
import org.slf4j.LoggerFactory;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.data.elasticsearch.core.query.NativeSearchQueryBuilder;
import org.springframework.stereotype.Service;

import com.bhtask.nhs.backend.BackendProperties;
import com.bhtask.nhs.backend.dao.Condition;
import com.bhtask.nhs.backend.dao.ConditionsRepository;
import com.fasterxml.jackson.databind.ObjectMapper;

@Service("elasticsearchConditionsServiceImpl")
public class ElasticsearchConditionsServiceImpl implements ConditionsService {

    private static final Logger logger = LoggerFactory.getLogger(ElasticsearchConditionsServiceImpl.class);

    @Autowired
    private BackendProperties backendProps;

    @Autowired
    private ConditionsRepository conditionsDao;

    @PostConstruct
    private void loadData() {
	final File dataFile = new File(backendProps.getDataPath());
	if (!dataFile.exists()) {
	    logger.error("Data file does not exist at path: {}.", backendProps.getDataPath());
	    return;
	}

	logger.info("Starting Elasticsearch index building from data file...");

	try (BufferedReader reader = new BufferedReader(new FileReader(dataFile))) {
	    ObjectMapper parser = new ObjectMapper();
	    String line = null;
	    while ((line = reader.readLine()) != null) {
		Condition condition = parser.readValue(line, Condition.class);
		logger.debug("indexing condition with URL: {}.", condition.getUrl());
		conditionsDao.save(condition);
	    }
	    logger.info("Finished loading data and building index.");
	} catch (Exception e) {
	    logger.error("Failed to index data file into Elasticsearch.", e);
	}
    }

    @Override
    public List<Condition> search(String query) {
	// @formatter:off
	QueryBuilder queryBuilder = new NativeSearchQueryBuilder()
		.withQuery(matchQuery(Condition.FIELD__CONTENT, query)
			.minimumShouldMatch("100%"))
		.build().getQuery();
	// @formatter:on

	final List<Condition> links = new ArrayList<>();
	conditionsDao.search(queryBuilder).forEach(links::add);
	return links;
    }

    @Override
    public List<String> getAllLinks() {
	// @formatter:off
	QueryBuilder queryBuilder = new NativeSearchQueryBuilder()
		.withQuery(matchAllQuery())
		.build().getQuery();
	// @formatter:on

	final List<String> links = new ArrayList<>();
	conditionsDao.search(queryBuilder).forEach(c -> links.add(c.getUrl()));
	return links;
    }

    @Override
    public long getCount() {
	return conditionsDao.count();
    }

}

File: nhs_backend/scala/src/main/scala/com/bhtask/nhs/backend/service/ElasticsearchConditionsServiceImpl.scala
package com.bhtask.nhs.backend.service

import java.lang

import com.bhtask.nhs.backend.BackendProperties
import com.bhtask.nhs.backend.dao.{Condition, ConditionsRepository}
import com.fasterxml.jackson.databind.ObjectMapper
import org.elasticsearch.index.query.QueryBuilder
import org.elasticsearch.index.query.QueryBuilders._
import org.slf4j.{Logger, LoggerFactory}
import org.springframework.beans.factory.annotation.Autowired
import org.springframework.data.elasticsearch.core.query.NativeSearchQueryBuilder
import org.springframework.stereotype.Service

import scala.collection.JavaConverters._
import scala.io.Source

@Service
class ElasticsearchConditionsServiceImpl(@Autowired val conditionsDao: ConditionsRepository,
                                         @Autowired val backendProps: BackendProperties)
  extends ConditionsService {

  private val logger: Logger = LoggerFactory.getLogger(classOf[ElasticsearchConditionsServiceImpl])

  override def search(query: String): Iterable[Condition] = {
    def queryBuilder: QueryBuilder = new NativeSearchQueryBuilder()
      .withQuery(matchQuery(Condition.FIELD__CONTENT, query)
        .minimumShouldMatch("100%"))
      .build().getQuery

    conditionsDao.search(queryBuilder).asScala
  }

  override def getAllLinks: Iterable[String] = conditionsDao.findAll().asScala.map(c => c.getTitle)

  override def getCount: Long = conditionsDao.count()

  override def loadIndex(): Boolean = {
    val dataFile = Source.fromFile(backendProps.data)
    try {
      val parser = new ObjectMapper

      logger info "Loading data file line-by-line into ES index."
      val conditions: lang.Iterable[Condition] = dataFile.getLines()
        .map(line => parser.readValue(line, classOf[Condition]))
        .toIterable.asJava

      conditionsDao.save(conditions)
      logger info "Finished data loading into ES index successfully."
      return true
    } catch {
      case t: Throwable => logger.error("Failed to load data file into index.", t)
    } finally {
      dataFile close
    }
    false
  }
}
